src/core/Tree/TreeNode.tsx

```diff
@@ -0,0 +1,199 @@
+  /**
+   * Checkbox to be shown before TreeNode.
+   * If undefined checkbox will not be shown.
+   * Recommended to use Checkbox component.
+   */
+  nodeCheckbox?: React.ReactNode;
+  /**
+   * Content shown after TreeNode.
+   */
+  children?: React.ReactNode;
+} & CommonProps;
+
+/**
+ * TreeNode component to display node content within a Tree.
+ * Must be used inside `Tree` component to correctly set node depth and subNodes.
+ * @example
+  <TreeNode
+    nodeId={props.nodeId}
+    label={props.node.label}
+    sublabel={props.node.sublabel}
+    onNodeExpanded={onNodeExpanded}
+    onNodeSelected={onSelectedNodeChange}
+    isDisabled={props.isDisabled}
+    isExpanded={props.isExpanded}
+    isSelected={props.isSelected}
+    nodeCheckbox={
+      <Checkbox variant='eyeball' disabled={props.isDisabled} />
+    }
+    icon={<SvgPlaceholder />}
+  />
+ */
+export const TreeNode = (props: TreeNodeProps) => {
+  const {
+    nodeId,
+    label,
+    sublabel,
+    children,
+    style,
+    className,
+    icon,
+    isDisabled = false,
+    isExpanded = false,
+    isSelected = false,
+    onNodeSelected,
+    onNodeExpanded,
+    nodeCheckbox,
+    ...rest
+  } = props;
+  useTheme();
+
+  const context = React.useContext(TreeContext);
+  const nodeDepth = context?.nodeDepth ?? 0;
+  const subNodeIds = context?.subNodeIds ?? [];
+
+  const styleDepth = React.useMemo(
+    () =>
+      getWindow()?.CSS?.supports?.(`--level: ${nodeDepth}`)
+        ? { '--level': nodeDepth, ...style }
+        : { marginLeft: nodeDepth ? nodeDepth * 28 : 0, ...style },
+    [nodeDepth, style],
+  );
+
+  const onNodeClick = () => {
+    if (isDisabled) {
+      return;
+    }
+    onNodeSelected?.(nodeId, !isSelected);
+  };
+
+  return (
+    <>
+      <li role='treeitem' id={nodeId} aria-expanded={isExpanded} {...rest}>
+        {
+          <div
+            className={cx(
+              'iui-tree-node',
+              {
+                'iui-active': isSelected,
+                'iui-disabled': isDisabled,
+              },
+              className,
+            )}
+            style={styleDepth}
+            onClick={onNodeClick}
+            tabIndex={isSelected ? 0 : -1}
+          >
+            {nodeCheckbox && React.isValidElement(nodeCheckbox)
+              ? React.cloneElement(nodeCheckbox, {
+                  className: cx(
+                    'iui-tree-node-checkbox',
+                    nodeCheckbox.props.className,
+                  ),
+                })
+              : undefined}
+            <div className='iui-tree-node-content'>
+              {context?.hasSubNodes && (
+                <IconButton
+                  styleType='borderless'
+                  size='small'
+                  onClick={(e) => {
+                    onNodeExpanded(nodeId, !isExpanded);
+                    e.stopPropagation();
+                  }}
+                  disabled={isDisabled}
```

Comment:
People are asking from time to time how can they enable expander or checkbox in Table when row is disabled. I see it might be the same situation. Solution could be by making TreeNode subcomponents for expander, icon and content. What do you think @mayank99 @veekeys

Reply:
I guess we need to think about it but not necessarily implement it now. Just have an option to extend if needed..
Or you think we need it now cause later might be considered as breaking change?

Reply:
I don't think we could do that in the future easily as we already show expander when `hasSubrows` is true and it also will be confusing that there is props and some subcomponents.

Reply:
> TreeNode subcomponents for expander, icon and content

Just expander and content should be enough because icon is really part of content

Reply:
Lets talk about this next week, what we could do.

Reply:
Added a prop for expander